[JENKINS-60381] Remove old, deprecated agent protocols.

jeffret-b opened the pull request:
Remove old, deprecated agent protocols that haven't been supported for a couple of years.

See [JENKINS-60381](https://issues.jenkins-ci.org/browse/JENKINS-60381).

This also includes the Remoting release 3.40. See the Remoting [changelog](https://github.com/jenkinsci/remoting/releases). This mostly involves removing the protocols from the Remoting library. It also includes a few other maintenance chores.

### Proposed changelog entries

* Remove old, deprecated, unsupported agent protocols Inbound TCP Agent Protocol/1, Inbound TCP Agent Protocol/2, and Inbound TCP Agent Protocol/3. ([issue 60381](https://issues.jenkins-ci.org/browse/JENKINS-60381))
* Update Remoting from 3.36 to 3.40 to remove unsupported protocols and minor maintenance improvements. (full [changelog](https://github.com/jenkinsci/remoting/releases))

### Proposed upgrade guidelines

==== Remove unsupported agent protocols
https://issues.jenkins-ci.org/browse/JENKINS-60381[JENKINS-60381]

Old, long deprecated, unsupported agent protocols have been removed from Jenkins. These are agent-initiated protocols, where the connection is launched on the agent. The protocols are currently called Inbound TCP Agent Protocol/1, Inbound TCP Agent Protocol/2, and Inbound TCP Agent Protocol/3. In other documents and times they have also been known as JNLP1-connect, JNLP2-connect, and JNLP3-connect.

The Inbound TCP Agent Protocol/4 (JNLP4-connect) was introduced in October 2016 to provide a more secure, stable, and scalable mechanism for agent-initiated connections. The prior protocols were deprecated and unsupported less than a year later as this protocol was proven superior. Since that time there has been an administrative monitor and UI warning against their use. This release removes the earlier protocols and the administrative monitor.

If you are still using an unsupported protocol you *must* manually upgrade to a supported connection mechanism. To continue using agent-initiated connections TCP Agent Protocol/4 is recommended. Agents using one of the unsupported protocols will not work until reconfigured.


### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
No tests needed for removed functionality.
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

jeffret-b committed:
Remove deprecated agent protocols.

oleg-nenashev commented:
Should we post an announcement in the jenkins.io blog ahead of the release
date? Maybe even an admin monitor?


On Thu, Dec 5, 2019, 17:04 Jeff Thompson <notifications@github.com> wrote:

> Remove old, deprecated agent protocols that haven't been supported for a
> couple of years.
>
> This should be paired with a new Remoting version containing
> jenkinsci/remoting#359 <https://github.com/jenkinsci/remoting/pull/359>,
> though it's not strictly required.
>
> See JENKINS-60381 <https://issues.jenkins-ci.org/browse/JENKINS-60381>.
> Proposed changelog entries
>
>    - Remove old, deprecated, unsupported agent protocols JNLP1-connect,
>    JNLP2-connect, and JNLP3-connect. (issue 60381
>    <https://issues.jenkins-ci.org/browse/JENKINS-60381>)
>
> Submitter checklist
>
>    - JIRA issue is well described
>    - Changelog entry appropriate for the audience affected by the change
>    (users or developer, depending on the change). Examples
>    <https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml>
>    * Use the Internal: prefix if the change has no user-visible impact
>    (API, test frameworks, etc.)
>    - Appropriate autotests or explanation to why this change has no tests
>    No tests needed for removed functionality.
>    - For dependency updates: links to external changelogs and, if
>    possible, full diffs
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/jenkinsci/jenkins/pull/4387
> Commit Summary
>
>    - Remove deprecated agent protocols.
>
> File Changes
>
>    - *M* core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-0> (2)
>    - *D*
>    core/src/main/java/jenkins/slaves/DeprecatedAgentProtocolMonitor.java
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-1> (94)
>    - *M* core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol.java
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-2> (40)
>    - *D* core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol2.java
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-3> (67)
>    - *D* core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol3.java
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-4> (64)
>    - *D*
>    core/src/main/resources/jenkins/slaves/DeprecatedAgentProtocolMonitor/message.jelly
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-5> (7)
>    - *D*
>    core/src/main/resources/jenkins/slaves/DeprecatedAgentProtocolMonitor/message.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-6> (5)
>    - *D*
>    core/src/main/resources/jenkins/slaves/DeprecatedAgentProtocolMonitor/message_bg.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-7> (35)
>    - *D*
>    core/src/main/resources/jenkins/slaves/DeprecatedAgentProtocolMonitor/message_it.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-8> (28)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol/deprecationCause.jelly
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-9> (4)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol/deprecationCause.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-10> (2)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol/deprecationCause_bg.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-11> (26)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol/deprecationCause_it.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-12> (24)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol/description.jelly
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-13> (4)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol/description.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-14> (2)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol/description_bg.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-15> (30)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol/description_de.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-16> (24)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol/description_it.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-17> (24)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol/description_sr.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-18> (3)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol2/deprecationCause.jelly
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-19> (5)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol2/deprecationCause.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-20> (3)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol2/deprecationCause_bg.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-21> (30)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol2/deprecationCause_it.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-22> (26)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol2/description.jelly
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-23> (4)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol2/description.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-24> (3)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol2/description_bg.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-25> (33)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol2/description_de.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-26> (23)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol2/description_it.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-27> (25)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol2/description_sr.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-28> (3)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol3/deprecationCause.jelly
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-29> (5)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol3/deprecationCause.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-30> (1)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol3/deprecationCause_bg.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-31> (26)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol3/deprecationCause_it.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-32> (24)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol3/description.jelly
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-33> (4)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol3/description.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-34> (1)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol3/description_bg.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-35> (35)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol3/description_de.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-36> (23)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol3/description_it.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-37> (23)
>    - *D*
>    core/src/main/resources/jenkins/slaves/JnlpSlaveAgentProtocol3/description_sr.properties
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-38> (3)
>    - *D* test/src/test/java/jenkins/AgentProtocolTest.java
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-39> (148)
>    - *M* test/src/test/java/jenkins/install/SetupWizardTest.java
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-40> (12)
>    - *M*
>    test/src/test/resources/hudson/model/QueueTest/load_queue_xml/config.xml
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-41> (4)
>    - *D*
>    test/src/test/resources/jenkins/AgentProtocolTest/testShouldNotDisableProtocolsForMigratedInstances/config.xml
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-42> (39)
>    - *D*
>    test/src/test/resources/jenkins/AgentProtocolTest/testShouldNotOverrideUserConfiguration/config.xml
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-43> (47)
>    - *M*
>    test/src/test/resources/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest/fromPreviousCustomSetup/config.xml
>    <https://github.com/jenkinsci/jenkins/pull/4387/files#diff-44> (4)
>
> Patch Links:
>
>    - https://github.com/jenkinsci/jenkins/pull/4387.patch
>    - https://github.com/jenkinsci/jenkins/pull/4387.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/4387?email_source=notifications&email_token=AAW4RIHF7SN7F5D4RRHHYI3QXEYCDA5CNFSM4JV4W6D2YY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4H6MH33A>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAW4RIFMVWTEPRKMDECJ6VTQXEYCDANCNFSM4JV4W6DQ>
> .
>

jeffret-b commented:
We should post on the blog about this. Not sure if it should be before or concurrent with making the change. What do you think?

I don't know what the admin monitor would do.

jvz reviewed (approved):
LGTM so far. I'll note that I tend to use SSH agents rather than JNLP, so I might not be the target audience, though.

jeffret-b commented:
Try the build again?

daniel-beck commented:
Build failed because `JnlpAgentReceiver.java` has Javadoc `@{link }`s to now removed classes.

daniel-beck commented on core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java:
```diff
@@ -216,5 +216,5 @@ public void setLog(@Nonnull OutputStream log) {
 
     private static final Logger LOGGER = Logger.getLogger(DefaultJnlpSlaveReceiver.class.getName());
 
-    private static final String COOKIE_NAME = JnlpSlaveAgentProtocol2.class.getName()+".cookie";
```
Given the old cookie name, is this only used for JNLP Agent Protocol/2? Or was it just introduced there, and is still used in later protocols?

jeffret-b replied:
It was introduced with an earlier protocol and use was continued, at least somewhat, with newer ones. Because of the way the cookie is roundtripped, the exact name doesn't matter so I cleaned it up a bit.

daniel-beck commented:
> Remove old, deprecated, unsupported agent protocols JNLP1-connect, JNLP2-connect, and JNLP3-connect

Please use the labels used on the UI for the changelog.

jeffret-b committed:
Cleanup JavaDoc from references to removed classes.

jeffret-b commented:
> Please use the labels used on the UI for the changelog.

I'm not sure what to call it. Bug fix? Enhancement? WDYT?

MarkEWaite commented:
> > Please use the labels used on the UI for the changelog.
> 
> I'm not sure what to call it. Bug fix? Enhancement? WDYT?

Some of the repositories that use release drafter have a `remove` label.  I think this qualifies as `remove`.  Unfortunately, this repository does not seem to have the `remove` label.

daniel-beck commented:
> I'm not sure what to call it. Bug fix? Enhancement? WDYT?

To clarify, I'm asking for "Inbound Agent Protocol/3" or whatever it is called in favor of "JNLP3-connect".

jeffret-b commented:
> To clarify, I'm asking for "Inbound Agent Protocol/3" or whatever it is called in favor of "JNLP3-connect".

Ah, that clarifies. I'll use the current naming.

jglick commented on core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol.java:
```diff
@@ -42,56 +42,20 @@
  */
-@Extension
-@Symbol("jnlp")
 public class JnlpSlaveAgentProtocol extends AgentProtocol {
```
Why not just delete it?

jglick replied:
But maybe that could be moved to `JnlpAgentReceiver`, the field here aliased to that one, and then this whole class deprecated?

jglick commented on core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol.java:
```diff
@@ -42,56 +42,20 @@
  */
-@Extension
-@Symbol("jnlp")
 public class JnlpSlaveAgentProtocol extends AgentProtocol {
```
```suggestion
public class JnlpSlaveAgentProtocol {
```

so you can get rid of the overrides.

jglick replied:
Also remove unused imports.

jglick replied:
(Set `remoting.version` to 3.40 to see that this is necessary.)

jeffret-b committed:
Update to Remoting without protocols.

jeffret-b committed:
Reduce JnlpSlaveAgentProtocol to minimum.

Remove anything unnecessary. Move SLAVE_SECRET to JnlpAgentReceiver
and alias it for existing usages.

jeffret-b commented:
Can anyone share information about the process for updating the upgrade guide? Do I just create a new file in a PR here: https://github.com/jenkins-infra/jenkins.io/tree/master/content/_data/upgrades ?

jeffret-b commented:
Thanks for reviewing, Jesse! I hadn't gotten around to finishing this one up after completing the Remoting side. I incorporated your suggestions so this should be ready to go once I get a good build.

oleg-nenashev reviewed (changes requested):
Requesting changes according to https://github.com/jenkinsci/jenkins/pull/4387#issuecomment-562283149 , the announcement is not ready and hence the PR cannot be shipped

jglick commented:
> We should post on the blog about this. Not sure if it should be before or concurrent with making the change.

IMO you can blog about this at any time—that should not be considered a blocker for merging. The weekly release notes are sufficient for people upgrading (who are unlikely to have stumbled across the blog post anyway).

jglick commented:
BTW I think `internal` is an inappropriate label here. Not sure what labels are honored here, though: https://github.com/jenkinsci/jenkins/blob/1f47ad6ea28090e3fd051dbed604e01d3dca707c/.github/release-drafter.yml#L28-L46

jglick commented:
Ah I see: https://github.com/jenkinsci/jenkins-core-changelog-generator/blob/b3e3ba3f397ae1d4cf50824158fff73c5f3fb0f8/generate-jenkins-changelog.rb#L64-L68

oleg-nenashev commented:
> IMO you can blog about this at any time—that should not be considered a blocker for merging. T

Blogpost is needed so that we can ensure that the removal and the potential downstream issues are properly documented for end users and plugin developers. I can live without a blogpost for now if this pull request includes explicit upgrade guidelines text draft we could reference from a weekly changelog

oleg-nenashev commented on test/src/test/java/jenkins/AgentProtocolTest.java:
```diff
@@ -1,148 +0,0 @@
- * 
- * @author Oleg Nenashev
- */
-public class AgentProtocolTest {
```
still valid, needs to be reworked to avoid usage of the removed code

oleg-nenashev reviewed (changes requested):
This test suite is still valid, there are other protocol implementations out there. E.g. in CloudBees Jenkins Operations Center ;) Removal of these tests reduces test coverage for the protocol management logic (e.g. the protocol deprecation engine), and IMO the tests should be retained. Hence I am requesting changes or an explicit justification why it is better to just remove these tests

jeffret-b commented:
I've been thinking about what to put in a blog post and I haven't come up with much beyond what I've just proposed for the upgrade guide. That has more than enough detail for what the user needs to do when upgrading, in a much more suitable location. A blog post might say something about the process involved in removing these protocols, or the issues. Maybe a historical review.

It might be interesting to create a combined blog post with information about the WebSockets Jesse is working on and some mention of removing these old protocols.

> Dev list announcement would be also nice

I've already discussed the removal on the dev list, though there was little interest in the topic. Once we're ready to go with this, I can follow up with a statement on that. Or would it be better to send that to the user list?

jeffret-b committed:
Add back in AgentProtocolTest.

Add back what was actually being used and remains.

oleg-nenashev commented:
@jeffret-b 

> I've already discussed the removal on the dev list, though there was little interest in the topic. Once we're ready to go with this, I can follow up with a statement on that. Or would it be better to send that to the user list?

Dev list is enough. Could you please post a link to the thread, I cannot find it

jeffret-b commented:
> Dev list is enough. Could you please post a link to the thread, I cannot find it

Oops, sorry. I remembered the wrong list. It was actually the user list. The findsecbugs discussion was dev list about the same time.

Try this link: https://groups.google.com/forum/#!topic/jenkinsci-users/s90MVVdFDyU

oleg-nenashev commented:
>  It was actually the user list

It explains a lot. I do not monitor it closely (shame on me). Kudos to @slide for doing so

slide commented:
The reason I thought it was ok to be on the user list is that might be the audience with the most people who would be impacted by the removal of agent protocols, the "users" of Jenkins. I can see where the Dev list would be appropriate too.

jeffret-b commented:
> The reason I thought it was ok to be on the user list is that might be the audience with the most people who would be impacted by the removal of agent protocols, the "users" of Jenkins. I can see where the Dev list would be appropriate too.

That's why I sent it to the User list. There are very few devs concerned with agent protocol implementation. Users who aren't concerned with keeping up are more likely to be interested. Of course, there may not be many of those on the User list either.

I'll send an announcement to both lists when we're prepared to move forward with this.

Wadeck reviewed (approved):
🐝 Did not look at the code but tested the PR locally and the protocol 4 still works as previously.

jeffret-b commented:
@oleg-nenashev or anyone else, what more do we need to do to get this moving forward? I don't see any outstanding issues. When we're ready to go, I'll announce the change in both the Dev and User lists. We'll see about putting together a blog post or other documentation afterwards.

oleg-nenashev reviewed (approved):
I think we are ready to go, thanks for the upgrade guidelines and the communication
I plan to merge it tomorrow if no negative feedback
